src/solver/lu_decomposition.cu

```diff
@@ -6,193 +6,174 @@
 #include <micm/util/cuda_param.hpp>
 #include <vector>
 
-// grouped parameters passing to DecomposeKernel()
-struct DecomposeDevice
-{
-  double* A_;
-  double* L_;
-  double* U_;
-  char* do_aik_;
-  size_t* aik_;
-  char* do_aki_;
-  size_t* aki_;
-  size_t* uii_;
-  std::pair<size_t, size_t>* niLU_;
-  std::pair<size_t, size_t>* uik_nkj_;
-  std::pair<size_t, size_t>* lij_ujk_;
-  std::pair<size_t, size_t>* lki_nkj_;
-  std::pair<size_t, size_t>* lkj_uji_;
-  size_t n_grids_;
-  size_t niLU_size_;
-};
-namespace micm
-{
-  namespace cuda
-  {
-    __global__ void DecomposeKernel(DecomposeDevice* device)
-    {
-      size_t tid = blockIdx.x * blockDim.x + threadIdx.x;
-      double* A = device->A_;
-      double* L = device->L_;
-      double* U = device->U_;
-      std::pair<size_t, size_t>* lkj_uji = device->lkj_uji_;
-      std::pair<size_t, size_t>* uik_nkj = device->uik_nkj_;
-      std::pair<size_t, size_t>* lij_ujk = device->lij_ujk_;
-      std::pair<size_t, size_t>* lki_nkj = device->lki_nkj_;
-      size_t do_aik_offset = 0;  // boolean vector
-      size_t aik_offset = 0;
-      size_t uik_nkj_offset = 0;
-      size_t lij_ujk_offset = 0;
-      size_t do_aki_offset = 0;  // boolean vector
-      size_t aki_offset = 0;
-      size_t lki_nkj_offset = 0;
-      size_t lkj_uji_offset = 0;
-      size_t uii_offset = 0;
-
-      if (tid < device->n_grids_)
-      {
-        // loop through every element in niLU
-        for (size_t i = 0; i < device->niLU_size_; i++)
+//grouped parameters passing to DecomposeKernel()
```

Comment:
@qinatan Ideally this PR only works on the LU solve code and should not touch other functions like LU decomposition. Could you explain more about why this code is revised?

Reply:
I passed size_t n_grids, size_t niLU_size to kernel function instead of grouping in the struct Device because referencing them through the pointer of Device cause memory segment fault.

Reply:
Thanks for the clarification.